node/src/reactor/participating.rs

```diff
@@ -616,12 +614,73 @@ impl reactor::Reactor for Reactor {
         // If not, run genesis or upgrade and construct a switch block, and use that for the latest
         // block header.
         let latest_block_header = if let Some(latest_block_header) = maybe_latest_block_header {
-            latest_block_header
+            if latest_block_header.protocol_version() == chainspec.protocol_config.version {
+                latest_block_header
+            } else if latest_block_header.protocol_version() < chainspec.protocol_config.version
+                && latest_block_header.is_switch_block()
+                && ActivationPoint::EraId(latest_block_header.next_block_era_id())
+                    == chainspec.protocol_config.activation_point
```

Comment:
Please consider extracting this 3-condition `if` into a named function (`is_block_last_one_before_upgrade()`?) and unit-test it.

Reply:
Done in https://github.com/casper-network/casper-node/pull/2416/commits/e8775595f89eda0208ca5223820b487f6c27bff7.